concent_api/api-test.py

```diff
@@ -1,72 +1,152 @@
+        task_id = task_id,
+    )
+
+    message_cannot_compute_task_dumped = dump(
+        message_cannot_compute_task,
+        CONCENT_PRIVATE_KEY,
+        CONCENT_PUBLIC_KEY,
+
+    )
+
+    return MessageRejectReportComputedTask(
+        timestamp = current_time,
+        reason = 'cannot-compute-task',
+        message_cannot_compute_task = message_cannot_compute_task_dumped,
+    )
+
+
+def print_golem_message(message, indent = 4):
+    assert isinstance(message, Message)
+
+    HEADER_FIELDS  = ['timestamp', 'encrypted', 'sig']
+    PRIVATE_FIELDS = {'_payload', '_raw'}
+
+    assert 'type' not in message.__slots__
+    fields = ['type'] + HEADER_FIELDS + sorted(set(message.__slots__) - set(HEADER_FIELDS) - PRIVATE_FIELDS)
+    values = [
+        type(message).__name__ if field == 'type' else
+        getattr(message, field)
+        for field in fields
+    ]
+
+    for field, value in zip(fields, values):
+        if isinstance(value, bytes):
+            try:
+                nested_message = load(
+                    value,
+                    CONCENT_PRIVATE_KEY,
+                    CONCENT_PUBLIC_KEY,
+                )
+            except AttributeError:
+                # FIXME: golem-messages provides no reliable way to discern invalid messages from other AttributeErrors.
+                print('{}{:30} = <BINARY DATA>'.format(' ' * indent, field))
+            else:
+                print('{}{:30} ='.format(' ' * indent, field))
+                print_golem_message(nested_message, indent = indent + 4)
+        else:
+            print('{}{:30} = {}'.format(' ' * indent, field, value))
 
 
 def api_request(host, endpoint, data = None, headers = None):
-    if headers is None:
-        headers = DEFAULT_HEADERS
+    assert endpoint not in ['', None]
+    assert host     not in ['', None]
+    assert headers  not in ['', None]
+
+    url = "{}/api/v1/{}/".format(host, endpoint)
 
     if data is None:
-        print('Receive message:')
+        print('RECEIVE ({})'.format(url))
     else:
-        print('Send message {}:'.format(data['type']))
-    print("{}/api/v1/{}/".format(host, endpoint))
+        print('SEND ({})'.format(url))
+        print('MESSAGE:')
+        print_golem_message(data)
+
+        data = dump(
+            data,
+            CONCENT_PRIVATE_KEY,
+            CONCENT_PUBLIC_KEY,
+        )
 
     if data is None:
-        response = requests.post("{}/api/v1/{}/".format(host, endpoint), headers = headers)
+        response = requests.post("{}".format(url), headers = headers)
     else:
-        response = requests.post("{}/api/v1/{}/".format(host, endpoint), headers = headers, data = json.dumps(data))
-
-    try:
-        decoded_response = response.json()
-    except json.JSONDecodeError:
-        decoded_response = response.text
-        print(response, decoded_response)
+        response = requests.post("{}".format(url), headers = headers, data = data)
+
+    if len(response.content) != 0:
```

Comment:
```if response.content:```
should be enough.

Reply:
We're usually explicit about checks like that.